File: simple_habex/script_simple_habex_model_test.m
% Copyright 2020, by the California Institute of Technology. ALL RIGHTS
% RESERVED. United States Government Sponsorship acknowledged. Any
% commercial use must be negotiated with the Office of Technology Transfer
% at the California Institute of Technology.
% -------------------------------------------------------------------------
% Script to test the Habex PROPER prescription in Matlab
%
% Written by A.J. Riggs (JPL, CIT) in February, 2020
% 
% Matlab's PATH must know the locations of 
% - the Habex prescription 'habex_vortex'
% - the PROPER library
% - the FALCO package

clear all

prescription = 'habex';

lambda0 = 550e-9;    %--Central wavelength of the whole spectral bandpass [meters]
lambda_um = lambda0 * 1e6;
optval.lambda0_um = 0.550;

%--Focal planes
res = 3;
FOV = 30;
Rout = 26; 
optval.nout = ceil_even(1 + res*(2*FOV)); %  dimensions of output in pixels (overrides output_dim0)
optval.final_sampling_lam0 = 1/res;	%   final sampling in lambda0/D
optval.use_field_stop = true;	%-- use field stop (0 = no stop)
optval.field_stop_radius = Rout;   %-- field stop radius in lam0/D

% %--Pupil Plane Resolutions
NbeamFull = 62*7;
optval.pupil_diam_pix = NbeamFull;
gridsize = 2^ceil(log2(NbeamFull)); 

optval.map_dir = '/Users/ajriggs/Documents/habex/maps/';	%-- directory containing optical surface error maps
optval.normLyotDiam = 0.95;
optval.vortexCharge = 6;

optval.xoffset = 0;
optval.yoffset = 0;

%% No Correction
%--Phase Retrieval Pupil

optval.use_pr = true;
optval.pr_pupil_diam_pix = 248;

optval.use_dm1 = false;
optval.use_errors = 1;		%-- 1 = use optical surface errors, 0 = none
optval.use_fpm = 0;		%-- use focal plane mask (0 = no FPM)
optval.use_lyot_stop = 1;	%-- use Lyot stop (0 = no stop)
optval.use_field_stop = 0;	%-- use field stop (0 = no stop)
% [Epup, sampling_m]  = habex_vortex(lambda_m, gridsize, optval);
[Epup, sampling_m] = prop_run(prescription, lambda_um, gridsize, 'quiet', 'passvalue', optval);

mask = 0*Epup;
mask(abs(Epup) > 0.1*max(abs(Epup(:)))) = 1;
% mask = ones(size(Epup));

figure(1); imagesc(abs(Epup)); axis xy equal tight; colorbar; title('abs(E$_{pupil}$)', 'Interpreter','Latex'); set(gca,'Fontsize',20); drawnow;
figure(2); imagesc(mask.*angle(Epup),[-1, 1]); axis xy equal tight; colorbar; title('angle(E$_{pupil}$)', 'Interpreter','Latex'); set(gca,'Fontsize',20); drawnow;

% PSF for normalization
optval.use_pr = false;
optval.use_errors = 1;		%-- 1 = use optical surface errors, 0 = none
optval.use_fpm = 0;		%-- use focal plane mask (0 = no FPM)
optval.use_lyot_stop = 1;	%-- use Lyot stop (0 = no stop)
optval.use_field_stop = 0;	%-- use field stop (0 = no stop)
% [EforNorm, sampling_m]  = habex_vortex(lambda_m, gridsize, optval);
[EforNorm, sampling_m] = prop_run(prescription, lambda_um, gridsize, 'quiet', 'passvalue', optval);

IforNorm = abs(EforNorm).^2;
I00 = max(IforNorm(:));
figure(3); imagesc(log10(IforNorm/I00),[-7 0]); axis xy equal tight; colorbar; title('PSF for Normalization)', 'Interpreter','Latex'); set(gca,'Fontsize',20); drawnow;

% Coronagraphic PSF
optval.use_errors = 1;		%-- 1 = use optical surface errors, 0 = none
optval.use_fpm = 1;		%-- use focal plane mask (0 = no FPM)
optval.use_lyot_stop = 1;	%-- use Lyot stop (0 = no stop)
optval.use_field_stop = 1;	%-- use field stop (0 = no stop)
% [Ecoro, sampling_m]  = habex_vortex(lambda_m, gridsize, optval);
[Ecoro, sampling_m] = prop_run(prescription, lambda_um, gridsize, 'quiet', 'passvalue', optval);
Inorm = abs(Ecoro).^2 / I00;
figure(4); imagesc(log10(Inorm), [-7 -2]); axis xy equal tight; colorbar;
title('Coronagraphic PSF before Flattening', 'Fontsize', 16); drawnow;


%% Corrected

optval.use_dm1 = true;
optval.dm1 = fitsread([optval.map_dir, 'flat_map.fits']);

%--Phase Retrieval Pupil
optval.use_pr = true;
optval.pr_pupil_diam_pix = 248;
optval.use_errors = 1;		%-- 1 = use optical surface errors, 0 = none
optval.use_fpm = 0;		%-- use focal plane mask (0 = no FPM)
optval.use_lyot_stop = 1;	%-- use Lyot stop (0 = no stop)
optval.use_field_stop = 0;	%-- use field stop (0 = no stop)
% [Epup, sampling_m]  = habex_vortex(lambda_um, gridsize, optval);
[Epup, sampling_m] = prop_run(prescription, lambda_um, gridsize, 'quiet', 'passvalue', optval);

mask = 0*Epup;
mask(abs(Epup) > 0.1*max(abs(Epup(:)))) = 1;
% mask = ones(size(Epup));

figure(11); imagesc(abs(Epup)); axis xy equal tight; colorbar; title('abs(E$_{pupil}$)', 'Interpreter','Latex'); set(gca,'Fontsize',20); drawnow;
figure(12); imagesc(mask.*angle(Epup),[-1, 1]); axis xy equal tight; colorbar; title('angle(E$_{pupil}$)', 'Interpreter','Latex'); set(gca,'Fontsize',20);drawnow;

% PSF for normalization
optval.use_pr = false;
optval.use_errors = 1;		%-- 1 = use optical surface errors, 0 = none
optval.use_fpm = 0;		%-- use focal plane mask (0 = no FPM)
optval.use_lyot_stop = 1;	%-- use Lyot stop (0 = no stop)
optval.use_field_stop = 0;	%-- use field stop (0 = no stop)
% [EforNorm, sampling_m]  = habex_vortex(lambda_m, gridsize, optval);
[EforNorm, sampling_m] = prop_run(prescription, lambda_um, gridsize, 'quiet', 'passvalue', optval);
IforNorm = abs(EforNorm).^2;
I00 = max(IforNorm(:));
figure(13); imagesc(log10(IforNorm/I00),[-7 0]); axis xy equal tight; colorbar; title('PSF for Normalization)', 'Interpreter','Latex'); set(gca,'Fontsize',20); drawnow;

% Coronagraphic PSF
optval.use_errors = 1;		%-- 1 = use optical surface errors, 0 = none
optval.use_fpm = 1;		%-- use focal plane mask (0 = no FPM)
optval.use_lyot_stop = 1;	%-- use Lyot stop (0 = no stop)
optval.use_field_stop = 1;	%-- use field stop (0 = no stop)
% [Ecoro, sampling_m]  = habex_vortex(lambda_m, gridsize, optval);
[Ecoro, sampling_m] = prop_run(prescription, lambda_um, gridsize, 'quiet', 'passvalue', optval);
Inorm = abs(Ecoro).^2 / I00;
figure(14); imagesc(log10(Inorm), [-7 -2]); axis xy equal tight; colorbar;
title('Coronagraphic PSF after Flattening', 'Fontsize', 16);
drawnow;

File: simple_habex/script_simple_habex_model_test.py
"""Show phase and amplitude at key planes in the Habex PROPER model."""
# Copyright 2020, by the California Institute of Technology. ALL RIGHTS
# RESERVED. United States Government Sponsorship acknowledged. Any
# commercial use must be negotiated with the Office of Technology Transfer
# at the California Institute of Technology.
# -------------------------------------------------------------------------
# Script to test the Habex PROPER prescription in Python
#
# Written by A.J. Riggs (JPL, CIT) in February, 2020
#
# PYTHONPATH (or sys.path) must know the locations of
# - the Habex prescription 'habex'
# - the PROPER library
# - the FALCO package

import numpy as np
from astropy.io import fits  # for reading in FITS files
import matplotlib.pyplot as plt
import proper
import sys
sys.path.insert(0, '/Users/ajriggs/Repos/proper-models/simple_habex')

map_dir = '/Users/ajriggs/Documents/habex/maps/'  # Change for your computer

prescription = 'habex'

# Focal Plane values
res = 3  # pixels per lambda0/D
FOV = 30  # FOV radius
Rout = 26  # field stop radius in lambda0/D

# Pupil Plane Resolutions
NbeamFull = 62*7  # points across the beam at the pupil
gridsize = int(2**np.ceil(np.log2(NbeamFull)))  # points across the pupil array

lambda0 = 550e-9  # Central wavelength of the whole spectral bandpass [meters]
lambda_um = 550e-9 * 1e6

optval = dict([
        ('map_dir', map_dir),
        ('nout', int(2*np.ceil(0.5*(1 + res*(2*FOV))))),
        ('final_sampling_lam0', 1/res),
        ('use_field_stop', True),
        ('field_stop_radius', Rout),
        ('lambda0_um', lambda0*1e6),
        ('pupil_diam_pix', NbeamFull),
        ('normLyotDiam', 0.95),
        ('vortexCharge', 6),
        ('xoffset', 0),
        ('yoffset', 0)
            ])

# %% No Correction

# Phase Retrieval of Pupil
optval['use_pr'] = True
optval['pr_pupil_diam_pix'] = 248
optval['use_errors'] = 1  # 1 = use optical surface errors, 0 = none
optval['use_fpm'] = 0  # use focal plane mask (0 = no FPM)
optval['use_lyot_stop'] = 1  # use Lyot stop (0 = no stop)
optval['use_field_stop'] = 0  # use field stop (0 = no stop)
[Epup, sampling_m] = proper.prop_run(prescription, lambda_um, gridsize,
                                     PASSVALUE=optval, QUIET=True)

mask = np.zeros(Epup.shape)
mask[np.abs(Epup) > 0.1*np.max(np.abs(Epup))] = 1

plt.figure(1); plt.imshow(np.abs(Epup)); plt.colorbar(); plt.title('abs(Epupil)'); plt.pause(1e-2)
plt.figure(2); plt.imshow(mask*np.angle(Epup)); plt.colorbar(); plt.clim(-1, 1); plt.title('angle(Epupil)'); plt.pause(1e-2)

# PSF for normalization
optval['use_pr'] = False
optval['use_errors'] = 1;		#-- 1 = use optical surface errors, 0 = none
optval['use_fpm'] = 0;		#-- use focal plane mask (0 = no FPM)
optval['use_lyot_stop'] = 1 	#-- use Lyot stop (0 = no stop)
optval['use_field_stop'] = 0 	#-- use field stop (0 = no stop)
[EforNorm, sampling_m] = proper.prop_run(prescription, lambda_um, gridsize, PASSVALUE=optval, QUIET=True )
IforNorm = np.abs(EforNorm)**2
I00 = np.max(IforNorm)
plt.figure(3); plt.imshow(np.log10(IforNorm/I00)); plt.colorbar(); plt.clim(-7, 0); plt.title('PSF for Normalization)'); plt.pause(1e-2)

# Coronagraphic PSF
optval['use_pr'] = False
optval['use_errors'] = 1;		#-- 1 = use optical surface errors, 0 = none
optval['use_fpm'] = 1;		#-- use focal plane mask (0 = no FPM)
optval['use_lyot_stop'] = 1;	#-- use Lyot stop (0 = no stop)
optval['use_field_stop'] = 1;#1;	#-- use field stop (0 = no stop)
[Ecoro, sampling_m] = proper.prop_run(prescription, lambda_um, gridsize, PASSVALUE=optval, QUIET=True )
Inorm = (np.abs(Ecoro)**2) / I00
plt.figure(4); plt.imshow(np.log10(Inorm)); plt.title('Aberrated Coronagraphic PSF'); plt.colorbar(); plt.clim(-7, -2); plt.pause(1)


# %% With Phase Flattened at Pupil before the FPM

optval['use_dm1'] = True
optval['dm1'] = fits.getdata(optval['map_dir']+'flat_map.fits')

#--Phase Retrieval of Pupil
optval['use_pr'] = True
optval['pr_pupil_diam_pix'] = 248;
optval['use_errors'] = 1;		#-- 1 = use optical surface errors, 0 = none
optval['use_fpm'] = 0;		#-- use focal plane mask (0 = no FPM)
optval['use_lyot_stop'] = 1;	#-- use Lyot stop (0 = no stop)
optval['use_field_stop'] = 0;	#-- use field stop (0 = no stop)
[Epup, sampling_m] = proper.prop_run(prescription, lambda_um, gridsize, PASSVALUE=optval, QUIET=True )

mask = np.zeros(Epup.shape)
mask[np.abs(Epup) > 0.1*np.max(np.abs(Epup))] = 1

plt.figure(11); plt.imshow(np.abs(Epup)); plt.colorbar(); plt.title('abs(Epupil)'); plt.pause(1e-2)
plt.figure(12); plt.imshow(mask*np.angle(Epup)); plt.colorbar(); plt.clim(-1, 1); plt.title('angle(Epupil)'); plt.pause(1e-2)


# PSF for normalization
optval['use_pr'] = False
optval['use_errors'] = 1;		#-- 1 = use optical surface errors, 0 = none
optval['use_fpm'] = 0;		#-- use focal plane mask (0 = no FPM)
optval['use_lyot_stop'] = 1 	#-- use Lyot stop (0 = no stop)
optval['use_field_stop'] = 0 	#-- use field stop (0 = no stop)
[EforNorm, sampling_m] = proper.prop_run(prescription, lambda_um, gridsize, PASSVALUE=optval, QUIET=True )
IforNorm = np.abs(EforNorm)**2
I00 = np.max(IforNorm)
plt.figure(13); plt.imshow(np.log10(IforNorm/I00)); plt.colorbar(); plt.clim(-7, 0); plt.title('PSF for Normalization)'); plt.pause(1e-2)


# Coronagraphic PSF
optval['use_pr'] = False
optval['use_errors'] = 1;		#-- 1 = use optical surface errors, 0 = none
optval['use_fpm'] = 1;		#-- use focal plane mask (0 = no FPM)
optval['use_lyot_stop'] = 1;	#-- use Lyot stop (0 = no stop)
optval['use_field_stop'] = 1;	#-- use field stop (0 = no stop)
[Ecoro, sampling_m] = proper.prop_run(prescription, lambda_um, gridsize, PASSVALUE=optval, QUIET=True )
Inorm = np.abs(Ecoro)**2 / I00
plt.figure(14); plt.imshow(np.log10(Inorm)); plt.title('Coronagraphic PSF after Flattening'); plt.clim(-7, -2); plt.colorbar(); plt.pause(5)
